File: libs/lifecycle/src/main/kotlin/mono/lifecycle/LifecycleOwner.kt
package mono.lifecycle

/**
 * An abstract class that has lifecycle from start (with [onStart]) to stop (with [onStop]).
 * The other classes can use this to self-aware lifecycle via [addObserver] method.
 */
abstract class LifecycleOwner {
    // Visible for testing only
    internal val lifecycleObservers: MutableList<LifecycleObserver> = mutableListOf()
    private var state: State = State.INITIAL

    fun addObserver(lifecycleObserver: LifecycleObserver) {
        if (state == State.STOPPED) {
            return
        }
        lifecycleObservers.add(lifecycleObserver)
        if (state == State.STARTED) {
            lifecycleObserver.onStart()
        }
    }

    fun onStart() {
        state = State.STARTED
        for (observer in lifecycleObservers) {
            observer.onStart()
        }
        onStartInternal()
    }

    protected open fun onStartInternal() = Unit

    fun onStop() {
        state = State.STOPPED
        onStopInternal()

        for (observer in lifecycleObservers) {
            observer.onStop()
        }
        lifecycleObservers.clear()
    }

    protected open fun onStopInternal() = Unit

    private enum class State {
        INITIAL, STARTED, STOPPED
    }
}

File: monosketch-svelte/src/lib/libs/flow/lifecycleowner.ts
export interface LifecycleObserver {
    onStart?(): void;

    onStop?(): void;
}

export enum LifecycleState {
    INITIALIZED,
    STARTED,
    STOPPED,
}

export class LifecycleOwner {
    private observers: LifecycleObserver[] = [];
    private state: LifecycleState = LifecycleState.INITIALIZED;

    static start(): LifecycleOwner {
        const owner = new LifecycleOwner();
        owner.onStart();
        return owner;
    }

    get isActive(): boolean {
        return this.state === LifecycleState.STARTED;
    }

    addObserver(observer: LifecycleObserver) {
        if (this.state === LifecycleState.STOPPED) {
            return;
        }

        this.observers.push(observer);
        if (this.state === LifecycleState.STARTED) {
            observer.onStart?.();
        }
    }

    /**
     * Called when the LifecycleOwner is ready to start.
     * Do not override this method, override onStartInternal instead.
     */
    onStart() {
        if (this.state === LifecycleState.STARTED) {
            return;
        }

        this.state = LifecycleState.STARTED;
        for (const observer of this.observers) {
            observer.onStart?.();
        }
        this.onStartInternal();
    }

    protected onStartInternal() {}

    /**
     * Called when the LifecycleOwner is ready to stop.
     * Do not override this method, override onStopInternal instead.
     */
    onStop() {
        if (this.state === LifecycleState.STOPPED) {
            return;
        }

        this.state = LifecycleState.STOPPED;
        this.onStopInternal();
        for (const observer of this.observers) {
            observer.onStop?.();
        }
    }

    protected onStopInternal() {}
}
